common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java

```diff
@@ -198,181 +173,446 @@
+            List<ExtensionDefinition> extensionDefinitions = new ArrayList<>();
+            try {
+                loadFile(SERVICES_DIRECTORY, loader, extensionDefinitions);
+                loadFile(SEATA_DIRECTORY, loader, extensionDefinitions);
+            } catch (IOException e) {
+                throw new EnhancedServiceNotFoundException(e);
             }
 
-            return Integer.compare(o1, o2);
-
-        });
+            //After loaded all the extensions,sort the caches by order
+            if (!nameToDefinitionsMap.isEmpty()) {
+                for (List<ExtensionDefinition> definitions : nameToDefinitionsMap.values()) {
+                    Collections.sort(definitions, (def1, def2) -> {
+                        int o1 = def1.getOrder();
+                        int o2 = def2.getOrder();
+                        return Integer.compare(o1, o2);
+                    });
+                }
+            }
 
-        return extensions;
-    }
+            if (!extensionDefinitions.isEmpty()) {
+                Collections.sort(extensionDefinitions, (definition1, definition2) -> {
+                    int o1 = definition1.getOrder();
+                    int o2 = definition2.getOrder();
+                    return Integer.compare(o1, o2);
+                });
+            }
 
-    @SuppressWarnings("rawtypes")
-    private static void loadFile(Class<?> service, String dir, ClassLoader classLoader, List<Class> extensions)
-        throws IOException {
-        String fileName = dir + service.getName();
-        Enumeration<URL> urls;
-        if (classLoader != null) {
-            urls = classLoader.getResources(fileName);
-        } else {
-            urls = ClassLoader.getSystemResources(fileName);
+            return extensionDefinitions;
         }
 
-        if (urls != null) {
-            while (urls.hasMoreElements()) {
-                java.net.URL url = urls.nextElement();
-                BufferedReader reader = null;
-                try {
-                    reader = new BufferedReader(new InputStreamReader(url.openStream(), Constants.DEFAULT_CHARSET));
-                    String line = null;
-                    while ((line = reader.readLine()) != null) {
-                        final int ci = line.indexOf('#');
-                        if (ci >= 0) {
-                            line = line.substring(0, ci);
-                        }
-                        line = line.trim();
-                        if (line.length() > 0) {
-                            try {
-                                extensions.add(Class.forName(line, true, classLoader));
-                            } catch (LinkageError | ClassNotFoundException e) {
-                                LOGGER.warn("load [{}] class fail. {}", line, e.getMessage());
+
+        @SuppressWarnings("rawtypes")
+        private void loadFile(String dir, ClassLoader classLoader, List<ExtensionDefinition> extensions)
```

Comment:
how about set？ What happens if I write two identical records in the spi file?

Reply:
Without adjusting the data structure, now I will check whether the ExtensionDefinition has been loaded . If it is the same two-line configuration, I will print a warn log and skip the repeated loading